analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs

```diff
@@ -0,0 +1,116 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class UseUnixEpochBase<TSyntaxKind, TLiteralExpression, TMemberAccessExpression> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+    where TLiteralExpression : SyntaxNode
+    where TMemberAccessExpression : SyntaxNode
+{
+    private const string DiagnosticId = "S6588";
+
+    protected override string MessageFormat => "Use \"{0}.UnixEpoch\" instead of creating {0} instances that point to the unix epoch time";
+
+    protected abstract bool IsDateTimeKindUtc(TMemberAccessExpression memberAccess);
+    protected abstract bool IsGregorianCalendar(SyntaxNode node);
+    protected abstract bool IsZeroTimeOffset(SyntaxNode node);
+
+    protected UseUnixEpochBase() : base(DiagnosticId) { }
+
+    protected sealed override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(start =>
+        {
+            if (!CompilationTargetsValidNetVersion(start.Compilation))
+            {
+                return;
+            }
+
+            context.RegisterNodeAction(
+                Language.GeneratedCodeRecognizer,
+                c =>
+                {
+                    var argumentCount = Language.Syntax.ArgumentExpressions(c.Node).Count();
+                    if (argumentCount < 3)
+                    {
+                        return;
+                    }
+
+                    var type = c.SemanticModel.GetTypeInfo(c.Node).Type;
```

Comment:
This will check every constructor call with 3 or more arguments with the semantic model. To improve the performance you can first check the name of the constructor call using only the syntax tree. If it's DateTime or DateTimeOffset then proceed with the semantic check. This will have some FNs:
- when `DateTime` or `DateTimeOffset` are aliased with the using directive
- when someone introduces a new custom type called `DateTime` or `DateTimeOffset` (you already covered it in the UTs)
Both of these scenarios are very rare, and the chance of any of them occurring when the developer also calls these constructors with these specific parameters is astronomical.
It's better to document these as FNs and make the analyzer more performant.